### PR TITLE
[4.x] Add command to bring the tenants up and down from maintenance and remove deprecated exception

### LINKED ISSUE #533: Disabling maintenance mode, and CLI commands for maintenance mode

### PR DESCRIPTION
**Pull request by stein-j**
This PR changes the maintenance mode to comply to the changes made in Laravel 8.

**New**
- Adds two CLI commands to the package: `tenancy:up` and `tenancy:down` with the same parameters as the default Laravel command except for the option `render` because it is not possible to render a page before Laravel is booted for a specific tenant.

- Adds the function `bringUpFromMaintenance` in the trait `MaintenanceMode`.

**Changes**

A tenant in maintenance mode now throws an `HttpException` exception instead of the deprecated `MaintenanceModeException` exception. 

Closes #533.

### CODE REVIEW

**Comment by stancl** [comment]
Hey. This looks good, thanks for the contribution! I think I had some maintenance mode related tasks written in my backlog for v4, so I'll probably merge this when working on the feature. Should be in January I think.

File: src/Commands/Down.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+     * The name and signature of the console command.
+     *
+     * @var string
+     */
+
+    protected $signature = 'tenancy:down
+                                 {--time= : The time when the app has been set to maintenance mode}
+                                 {--message= : Message to display}
+                                 {--retry= : The number of seconds after which the request may be retried}
+                                 {--allowed=* : List of IPs allowed}';
```
Comment (stancl) [inline src/Commands/Down.php]: This doesn't implement the same parameters as the default `artisan` down command, right?

For instance, the `secret` — https://github.com/laravel/framework/blob/8.x/src/Illuminate/Foundation/Console/DownCommand.php#L23

**Comment by stancl** [comment]
Closing #757 and will finish that here. We could probably add `--secret` but I'm not sure if there's a point in having a different secret than the central app.

**Comment by stein-j** [comment]
Laravel now simply throws a `HttpException` exception. https://github.com/laravel/framework/blob/8f3b411969ffdd733a0ba54098e08f8ca7c1369e/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php#L78

If you give me a few days, I can update the code and comply to the new maintenance mode used by Laravel and then update the commands so that we can use the same parameters as [Laravel](https://github.com/laravel/framework/blob/5398fbbbf81ed842cf0430b278f07eb0f869cf8f/src/Illuminate/Foundation/Console/DownCommand.php#L19).
I think all parameters can be recreated except for `render` because this allows to return a response even before Laravel is initialised (which means, before the package is also initialised).

Note: Naturally this would allow #757

**Comment by stancl** [comment]
Sure, thanks a lot for the work on this! FWIW if there's a signature change, you can use the version in Laravel 9. And probably change the target to the `master` branch.

**Comment by stein-j** [comment]
I'm glade you mentioned Laravel 9 because coincidently two PRs ([1](https://github.com/laravel/framework/pull/40070), [2](https://github.com/laravel/framework/pull/40102)) have been accepted during those last few days and it changed the maintenance system.

L9 will now have a configurable driver for the maintenance storage system (which will determine how and where the data is stored), at the time of writing this comment there are two drivers/managers **file**: Which is the old way to put the application to maintenance ; **cache** : Which will store the data inside the cache.

The reason they added drivers, is because when using a load balancer and you set the application into maintenance mode it would only work on the application the command has been typed on... so a _cache_ driver was needed. But this package doesn't need such implementation because even if you would use a load balancer, there should be only one database connection **per tenant** for all applications (at least in most use cases).

So I think to keep things simple, this PR will keep its own feature and keep storing the data in the database. It might be important to mention in the documentation that the maintenance mode doesn't uses the driver from the config file.

Maybe in another PR we can try to create our custom driver, tenant compatible.

**Comment by stancl** [comment]
> this PR will keep its own feature and keep storing the data in the database

That is in the `data` column in our case, right?

**Comment by stein-j** [comment]
Yes

**Comment by stancl** [comment]
Just a note: the tests will fail because `master` has some failing tests due to dependencies that need new syntax. But we can look at this feature's tests in isolation

**Comment by stein-j** [comment]
@stancl the PR is done and all the tests have been added.
As you said there are conflict versions with composer, so the tests fails. Do you want me to look into it ? Or should we wait that the master to be updated ?

**Comment by stancl** [comment]
Probably wait until master is updated 👍🏻 

When we add L9 support, I'll merge 3.x into master

**Comment by stein-j** [comment]
Hello @stancl,

I just realized that the changes from this PR are not compatible with Laravel 6 & 7. I've put this PR back to draft, ping me when you start working on the V4 and I'll see what can be done at that point.

**Comment by stancl** [comment]
@lukinovec I think we should finish this PR before the pending tenants since it's smaller and it interacts with the trait adding a `--tenants` option. That way we'll make the conflict resolution simpler.

Also, the tests are showing as failing, but I think they'll run differently when the CI is triggered again since I simplified the CI matrix in the `master` branch. So if you could resolve the conflicts, that would trigger a CI build and we'll see how the feature works now.

**Comment by stancl** [comment]
Only php-cs-fixer is failing, tests seem to be passing, so CI-wise this PR is ready now.

I'll look at the code now and see if this is ready to be merged.

File: src/Commands/Down.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    protected $description = 'Put tenants into maintenance';
+
+    /**
+     * Execute the console command.
+     *
+     * @return mixed
+     */
+    public function handle()
+    {
+        tenancy()->runForMultiple($this->option('tenants'), function ($tenant) {
```
Comment (stancl) [inline src/Commands/Down.php]: @lukinovec Does the `HasATenantsOption` trait *require* `--tenants` or make it optional?

If it's optional, then it's possible that `$this->option('tenants')` will return `null`. Which would run this for **all** tenants.

Which can be the desired behavior, and probably is the desired behavior, but it's not immediately clear from looking at the code. So if you can confirm what I wrote above, we probably want to add a comment like this:
```suggestion
    {
        // This runs for all tenants if no --tenants are specified
        tenancy()->runForMultiple($this->option('tenants'), function ($tenant) {
```

Comment (stancl) [reply src/Commands/Down.php]: And you can probably do the same in `Up.php`

**Comment by stancl** [comment]
@stein-j Is anything unimplemented yet or is the PR fully ready? Is the code made to work with Laravel 9 (the changes you made to the way tenants are put into maintenance mode)?

**Comment by stein-j** [comment]
@stancl I want to go over it again, it's been a while and a file was changed in another PR since (https://github.com/archtechx/tenancy/pull/802/files#diff-40487242039e9842dc87ba96285282e42d0cc5b9a7dbeac6f115c4ee2bb8db59). Let me get back to you in a few hours.

**Comment by stein-j** [comment]
@stancl what will be the minimum requirements for Laravel and PHP for V4 ?

**Comment by stancl** [comment]
Just that the syntax is updated for Laravel 9. I *think* you already did this since the PR is recent. But I'm just checking in case anything is missing.

**Comment by stein-j** [comment]
Ok, it's all good.

File: src/Commands/Down.php (hunk 2 above)
Comment (lukinovec) [reply src/Commands/Down.php]: Yeah, the option is optional, and if not specified, the code runs for all tenants

Comment (lukinovec) [reply src/Commands/Down.php]: I added the comments

**Comment by stancl** [comment]
FYI I merged `master` into this so you should delete the `3.x` branch after we merge this since it isn't compatible with our `3.x` (in case you plan on making PRs to `3.x` in the future)

**Comment by stancl** [comment]
Thanks for your work on this PR!

**Pull request merged**